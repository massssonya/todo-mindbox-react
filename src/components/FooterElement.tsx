import {
	removeCompletedTodos,
	selectNumberActiveTodos,
	setSelector
} from "../redux/slices/todosSlice";
import { useAppDispatch, useAppSelector } from "../redux/hooks";

export const FooterElement = () => {
	const numberActiveTodos = useAppSelector(selectNumberActiveTodos);
	const dispatch = useAppDispatch();
	return (
		<footer className="px-4 py-2 flex flex-row items-center justify-between text-base border">
			<span>{numberActiveTodos} items left</span>
			<div className="flex flex-row gap-2 items-center justify-between">
				<button
					id="all"
					onClick={() => dispatch(setSelector("selectAllTodos"))}
				>
					All
				</button>
				<button
					id="active"
					onClick={() => dispatch(setSelector("selectActiveTodos"))}
				>
					Active
				</button>
				<button
					id="completed"
					onClick={() => dispatch(setSelector("selectCompletedTodos"))}
				>
					Completed
				</button>
			</div>
			<button onClick={() => dispatch(removeCompletedTodos())}>
				Clear Completes
			</button>
		</footer>
	);
};
