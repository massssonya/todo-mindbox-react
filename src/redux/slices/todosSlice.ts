import { createSlice } from "@reduxjs/toolkit";
import type { PayloadAction } from "@reduxjs/toolkit";

interface ITodo {
	id: string;
	value: string;
	isCompleted: boolean;
}

interface IState {
	todos: ITodo[];
	selector: string;
}

export const initialState: IState  = {
	todos: [
		{
			id: "todo1",
			value: "Task 1",
			isCompleted: false
		},
		{
			id: "todo2",
			value: "Task 2",
			isCompleted: true
		},
		{
			id: "todo3",
			value: "Task 3",
			isCompleted: false
		}
	],
	selector: "selectAllTodos"
};

export const todosSlice = createSlice({
	name: "todos",
	initialState,
	selectors: {
		selectAllTodos: (state) => state.todos,
		selectActiveTodos: (state) =>
			state.todos.filter((todo) => !todo.isCompleted),
		selectCompletedTodos: (state) =>
			state.todos.filter((todo) => todo.isCompleted),
		selectNumberActiveTodos: (state) =>
			state.todos.filter((todo) => !todo.isCompleted).length,
		selectSelector: (state) => state.selector
	},
	reducers: {
		addTodo: (state, action: PayloadAction<ITodo["value"]>) => {
			const newId = Date.now().toString();
			state.todos.push({
				id: newId,
				value: action.payload,
				isCompleted: false
			});
		},
		toggleTodo: (state, action: PayloadAction<ITodo["id"]>) => {
			return {
				...state,
				todos: state.todos.map((todo) => {
					if (todo.id === action.payload) {
						return { ...todo, isCompleted: !todo.isCompleted };
					}
					return todo;
				})
			};
		},
		removeCompletedTodos: (state) => {
			return {
				...state,
				todos: state.todos.filter((todo) => !todo.isCompleted)
			};
		},
		setSelector: (state, action: PayloadAction<IState["selector"]>) => {
			state.selector = action.payload;
		}
	}
});

export const { addTodo, toggleTodo, removeCompletedTodos, setSelector } =
	todosSlice.actions;
export const {
	selectAllTodos,
	selectNumberActiveTodos,
	selectActiveTodos,
	selectCompletedTodos,
	selectSelector
} = todosSlice.selectors;

export default todosSlice.reducer;
