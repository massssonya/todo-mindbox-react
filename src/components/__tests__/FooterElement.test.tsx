import { render, screen, fireEvent, waitFor } from "@testing-library/react";
import "@testing-library/jest-dom";
import * as reduxHooks from "../../redux/hooks";
import { FooterElement } from "../FooterElement";
import * as actions from "../../redux/slices/todosSlice";

jest.mock("../../redux/hooks");

const initialState = {
	todos: [
		{ id: "1", value: "test", isCompleted: false },
		{ id: "2", value: "test2", isCompleted: true }
	],
	selector: "selectAllTodos"
};

const useAppDispatch = jest.spyOn(reduxHooks, "useAppDispatch");
const useAppSelector = jest.spyOn(reduxHooks, "useAppSelector");
const selectNumberActiveTodos = jest.fn(
	(state: actions.IState) =>
		state.todos.filter((todo) => !todo.isCompleted).length
);

describe("FooterElement", () => {
	it("should render correctly", () => {
		useAppSelector.mockReturnValue(selectNumberActiveTodos(initialState));
		const component = render(<FooterElement />);
		expect(component).toMatchSnapshot();
	});

	it("should render number of active todos", () => {
		useAppSelector.mockReturnValue(selectNumberActiveTodos(initialState));
		render(<FooterElement />);
		expect(screen.getByText(/1/)).toBeInTheDocument();
	});

	it("should return zero if click on Clear Completes", () => {
		useAppSelector.mockReturnValue(selectNumberActiveTodos(initialState));
		const dispatch = jest.fn();
		useAppDispatch.mockReturnValue(dispatch);
		render(<FooterElement />);
		const clearCompleted = screen.getByText(/Clear Completes/);
		fireEvent.click(clearCompleted);
		expect(dispatch).toHaveBeenCalledWith(actions.removeCompletedTodos());
		expect(dispatch).toHaveBeenCalledTimes(1);
		waitFor(() => {
			expect(screen.getByText(/0/)).toBeInTheDocument();
		});
	});
});
