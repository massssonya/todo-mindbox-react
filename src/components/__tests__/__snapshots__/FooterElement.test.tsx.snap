// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FooterElement should render correctly 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <footer
        class="flex flex-row items-center justify-between text-base"
      >
        <span>
          1
           items left
        </span>
        <div
          class="flex flex-row gap-2 items-center justify-between"
        >
          <button
            id="all"
          >
            All
          </button>
          <button
            id="active"
          >
            Active
          </button>
          <button
            id="completed"
          >
            Completed
          </button>
        </div>
        <button>
          Clear Completes
        </button>
      </footer>
    </div>
  </body>,
  "container": <div>
    <footer
      class="flex flex-row items-center justify-between text-base"
    >
      <span>
        1
         items left
      </span>
      <div
        class="flex flex-row gap-2 items-center justify-between"
      >
        <button
          id="all"
        >
          All
        </button>
        <button
          id="active"
        >
          Active
        </button>
        <button
          id="completed"
        >
          Completed
        </button>
      </div>
      <button>
        Clear Completes
      </button>
    </footer>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
